balancer/weightedroundrobin/weightedroundrobin.go

```diff
@@ -27,3 +39,169 @@ const Name = "weighted_round_robin"
+		weight := uint32(1)
+		if addrInfo, ok := a.Metadata.(*AddrInfo); ok {
+			weight = addrInfo.Weight
+		}
+		addrWeights[a.Addr] += weight
+		if _, ok := b.subConns[a.Addr]; !ok {
+			// a is a new address (not existing in b.subConns).
+			sc, err := b.cc.NewSubConn([]resolver.Address{a}, balancer.NewSubConnOptions{HealthCheckEnabled: true})
+			if err != nil {
+				grpclog.Warningf("base.baseBalancer: failed to create new SubConn: %v", err)
+				continue
+			}
+			balancerAddrInfo := balancerAddrInfo{SubConn: sc}
+			b.subConns[a.Addr] = &balancerAddrInfo
+			b.scStates[sc] = connectivity.Idle
+			sc.Connect()
+		}
+	}
+	for a, sc := range b.subConns {
+		// a was removed by resolver.
+		if weight, ok := addrWeights[a]; !ok {
+			b.cc.RemoveSubConn(sc.SubConn)
+			delete(b.subConns, a)
+			// Keep the state of this sc in b.scStates until sc's state becomes Shutdown.
+			// The entry will be deleted in HandleSubConnStateChange.
+		} else {
+			sc.Weight = weight
+		}
+	}
+}
+
+func (b *wrrBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	s := state.ConnectivityState
+	grpclog.Infof("wrr: handle SubConn state change: %p, %v", sc, s)
+	oldS, ok := b.scStates[sc]
+	if !ok {
+		grpclog.Infof("wrr: got state changes for an unknown SubConn: %p, %v", sc, s)
+		return
+	}
+	b.scStates[sc] = s
+	switch s {
+	case connectivity.Idle:
+		sc.Connect()
+	case connectivity.Shutdown:
+		// When an address was removed by resolver, b called RemoveSubConn but
+		// kept the sc's state in scStates. Remove state for this sc here.
+		delete(b.scStates, sc)
+	}
+
+	oldAggrState := b.state
+	b.state = b.csEvltr.RecordTransition(oldS, s)
+
+	// Regenerate picker when one of the following happens:
+	//  - this sc became ready from not-ready
+	//  - this sc became not-ready from ready
+	//  - the aggregated state of balancer became TransientFailure from non-TransientFailure
+	//  - the aggregated state of balancer became non-TransientFailure from TransientFailure
+	if (s == connectivity.Ready) != (oldS == connectivity.Ready) ||
+		(b.state == connectivity.TransientFailure) != (oldAggrState == connectivity.TransientFailure) {
+		b.regeneratePicker()
+	}
+
+	b.cc.UpdateBalancerState(b.state, b.picker)
+}
+
+// regeneratePicker takes a snapshot of the balancer, and generates a picker
+// from it. The picker is
+//  - errPicker with ErrTransientFailure if the balancer is in TransientFailure,
+//  - built by the pickerBuilder with all READY SubConns otherwise.
+func (b *wrrBalancer) regeneratePicker() {
+	if b.state == connectivity.TransientFailure {
+		b.picker = base.NewErrPicker(balancer.ErrTransientFailure)
+		return
+	}
+	readySCs := make(map[string]*balancerAddrInfo)
+
+	// Filter out all ready SCs from full subConn map.
+	for addr, sc := range b.subConns {
+		if st, ok := b.scStates[sc.SubConn]; ok && st == connectivity.Ready {
+			readySCs[addr] = sc
+		}
+	}
+	b.picker = buildPicker(readySCs)
```

Comment:
EDF is stateful. This will reset its state whenever a connection changes between ready/not-ready.

Do we want to update the picker to only remove/add the subConn, but keeping the state of other subConns?

Reply:
Current WRR implementation doesn't allow modifications other than Add. Going to add MutableWRR interface with Update/Remove functionality.

Reply:
Actually we can do it without new interface assuming "Add=Update" and "Add(weight=0)=Remove"

Reply:
Actually the wrr interface is internal. So we can change it.

Reply:
Problem here is that adding mutations to the interface will require RandomWRR to follow. Update/Delete will be linear in RandomWRR.

Reply:
Updated implementation. I don't like the result, but may be it is good enough for now.